File: ios/Consent.swift
import ExpoModulesCore
import UserMessagingPlatform

/// Maps a UMP error onto the JS-facing `ConsentErrorCode`.
///
/// The numeric codes cannot be sent to JS as-is: `UMPErrorDomain` numbers `2` as
/// `invalidAppID`, while Android's `FormError.ErrorCode` numbers `2` as `INTERNET_ERROR`. The
/// same number would mean two different failures depending on the platform, so it is normalized
/// here and the raw number is appended to the message for diagnostics instead.
///
/// `UMPRequestErrorCode` (1...4) and `UMPFormErrorCode` (5...9) share one domain, so a single
/// switch covers both.
func consentErrorCode(_ error: NSError) -> String {
  guard error.domain == UMPErrorDomain else { return "unknown" }
  switch error.code {
  case 1, 5: return "internal"           // request internal, form internal
  case 2, 4: return "misconfiguration"   // invalidAppID, misconfiguration
  case 3: return "network"
  case 6, 9: return "invalidOperation"   // formAlreadyUsed, invalidViewController
  case 7: return "formUnavailable"
  case 8: return "timeout"
  default: return "unknown"
  }
}

func consentErrorMessage(_ error: NSError) -> String {
  "\(error.localizedDescription) (native code: \(error.code))"
}

/// Every UMP entry point is main-thread-only.
///
/// `UMPConsentInformation.h` states "Consent information. All methods must be called on the main
/// thread" for the whole class — property getters included — and every `UMPConsentForm` class
/// method carries "Must be called on the main queue".
///
/// This is why the JS API is async all the way down, including `getConsentInfo()`. The
/// alternative — a synchronous `Function` hopping to main with `DispatchQueue.main.sync` — is the
/// deadlock shape documented on `runOnMain` in `BannerAd.swift`, and unlike the banner size
/// helpers there is no requirement here that forces a synchronous answer.
@MainActor
enum Consent {
  static func snapshot() -> [String: Any] {
    let info = ConsentInformation.shared
    return [
      "status": statusString(info.consentStatus),
      "canRequestAds": info.canRequestAds,
      // Narrowed from iOS's three-valued `formStatus` to the boolean Android reports.
      // `.unknown` becomes false: an app can only branch on "can a form be shown", and Android
      // has no way to produce a third value. See `ConsentInfo.isConsentFormAvailable` in types.ts.
      "isConsentFormAvailable": info.formStatus == .available,
      "privacyOptionsRequirement": privacyString(info.privacyOptionsRequirementStatus),
    ]
  }

  static func requestInfoUpdate(_ options: [String: Any?]?) async throws {
    let parameters = makeRequestParameters(options)
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      ConsentInformation.shared.requestConsentInfoUpdate(with: parameters) { error in
        if let error { continuation.resume(throwing: error) } else { continuation.resume() }
      }
    }
  }

  static func showFormIfRequired(from viewController: UIViewController?) async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      ConsentForm.loadAndPresentIfRequired(from: viewController) { error in
        if let error { continuation.resume(throwing: error) } else { continuation.resume() }
      }
    }
  }

  static func showPrivacyOptionsForm(from viewController: UIViewController?) async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      ConsentForm.presentPrivacyOptionsForm(from: viewController) { error in
        if let error { continuation.resume(throwing: error) } else { continuation.resume() }
      }
    }
  }

  static func reset() {
    ConsentInformation.shared.reset()
  }

  // MARK: - Conversions

  private static func statusString(_ status: ConsentStatus) -> String {
    switch status {
    case .required: return "required"
    case .notRequired: return "notRequired"
    case .obtained: return "obtained"
    default: return "unknown"
    }
  }

  private static func privacyString(_ status: PrivacyOptionsRequirementStatus) -> String {
    switch status {
    case .required: return "required"
    case .notRequired: return "notRequired"
    default: return "unknown"
    }
  }

  private static func makeRequestParameters(_ options: [String: Any?]?) -> RequestParameters {
    let parameters = RequestParameters()
    guard let options else { return parameters }

    if let underAge = options["tagForUnderAgeOfConsent"] as? Bool {
      parameters.isTaggedForUnderAgeOfConsent = underAge
    }

    if let debug = options["debugSettings"] as? [String: Any?] {
      let debugSettings = DebugSettings()
      if let testDeviceIds = debug["testDeviceIds"] as? [String] {
        debugSettings.testDeviceIdentifiers = testDeviceIds
      }
      if let geography = debug["geography"] as? String {
        switch geography {
        case "eea": debugSettings.geography = .EEA
        case "regulatedUsState": debugSettings.geography = .regulatedUSState
        case "other": debugSettings.geography = .other
        // `.notEEA` (2) is deprecated in favour of `.other` (4) and is deliberately not
        // reachable from JS — `DebugGeography` in types.ts has no member for it.
        default: debugSettings.geography = .disabled
        }
      }
      parameters.debugSettings = debugSettings
    }

    return parameters
  }
}

File: android/src/main/java/expo/modules/googlemobileads/Consent.kt
package expo.modules.googlemobileads

import android.app.Activity
import android.content.Context
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.FormError
import com.google.android.ump.UserMessagingPlatform

/**
 * Maps a UMP [FormError] onto the JS-facing `ConsentErrorCode`.
 *
 * The numeric codes cannot be sent to JS as-is: Android's `FormError.ErrorCode` numbers `2` as
 * `INTERNET_ERROR`, while iOS's `UMPErrorDomain` numbers `2` as `invalidAppID`. The same number
 * would mean two different failures depending on the platform.
 *
 * Android's set is much smaller than iOS's — there is no equivalent of `misconfiguration` or
 * `formUnavailable`, which is documented on `ConsentErrorCode` in types.ts.
 */
fun consentErrorCode(code: Int): String = when (code) {
  1 -> "internal"          // INTERNAL_ERROR
  2 -> "network"           // INTERNET_ERROR
  3 -> "invalidOperation"  // INVALID_OPERATION
  4 -> "timeout"           // TIME_OUT
  else -> "unknown"
}

fun consentErrorMessage(error: FormError): String =
  "${error.message} (native code: ${error.errorCode})"

fun consentInformation(context: Context): ConsentInformation =
  UserMessagingPlatform.getConsentInformation(context)

fun consentSnapshot(context: Context): Map<String, Any?> {
  val info = consentInformation(context)
  return mapOf(
    "status" to when (info.consentStatus) {
      ConsentInformation.ConsentStatus.REQUIRED -> "required"
      ConsentInformation.ConsentStatus.NOT_REQUIRED -> "notRequired"
      ConsentInformation.ConsentStatus.OBTAINED -> "obtained"
      else -> "unknown"
    },
    "canRequestAds" to info.canRequestAds(),
    // iOS narrows its three-valued formStatus to this same boolean. See
    // `ConsentInfo.isConsentFormAvailable` in types.ts.
    "isConsentFormAvailable" to info.isConsentFormAvailable,
    "privacyOptionsRequirement" to when (info.privacyOptionsRequirementStatus) {
      ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED -> "required"
      ConsentInformation.PrivacyOptionsRequirementStatus.NOT_REQUIRED -> "notRequired"
      else -> "unknown"
    }
  )
}

fun makeConsentRequestParameters(
  activity: Activity,
  options: Map<String, Any?>?
): ConsentRequestParameters {
  val builder = ConsentRequestParameters.Builder()
  if (options == null) return builder.build()

  (options["tagForUnderAgeOfConsent"] as? Boolean)?.let(builder::setTagForUnderAgeOfConsent)

  (options["debugSettings"] as? Map<*, *>)?.let { debug ->
    val debugBuilder = ConsentDebugSettings.Builder(activity)
    (debug["testDeviceIds"] as? List<*>)?.filterIsInstance<String>()?.forEach {
      debugBuilder.addTestDeviceHashedId(it)
    }
    (debug["geography"] as? String)?.let { geography ->
      debugBuilder.setDebugGeography(
        when (geography) {
          "eea" -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA
          "regulatedUsState" ->
            ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_REGULATED_US_STATE
          "other" -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_OTHER
          // DEBUG_GEOGRAPHY_NOT_EEA is deprecated in favour of OTHER and is deliberately not
          // reachable from JS — `DebugGeography` in types.ts has no member for it.
          else -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_DISABLED
        }
      )
    }
    // `setForceTesting` is deliberately not exposed: iOS has no equivalent, and an emulator is
    // already treated as a test device, so QA does not need it.
    builder.setConsentDebugSettings(debugBuilder.build())
  }

  // `setAdMobAppId` is deliberately not exposed either: iOS has no equivalent, and the app ID is
  // already injected into AndroidManifest.xml by the config plugin.
  return builder.build()
}
